mansrc/curve2hints.1
'\" t
.\"     Title: curve2hints
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "CURVE2HINTS" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
curve2hints \- tool to make exon segmentation and grouping the exons together to form AUGUSTUS hints
.SH "SYNOPSIS"
.sp
\fBcurve2hints\fP \-\-in=file \-\-outexons=out.gff \-\-c=int \-\-s=float
.SH "OPTIONS"
.sp
\fB\-\-in=s\fP
.RS 4
iput file is single file containing tracks in wig format
.RE
.sp
\fB\-\-outexons=s\fP
.RS 4
file for the output segmentaion
.RE
.sp
\fB\-\-c\fP
.RS 4
specify the threshold to distinguish between introns and exons in terms of coverage depths (default 4)
.RE
.sp
\fB\-\-s\fP
.RS 4
specify the ratio \fB(1\-lambda1)/(1\-lambda2)\fP It should be >1 (default 1.12)
.RE
.sp
\fB\-\-z\fP
.RS 4
specify the chunk size (default value 50000)
.RE
.sp
\fB\-\-r\fP
.RS 4
To enable re\-estimation of distribution from the output produced
.RE
.sp
\fB\-\-train_file\fP
.RS 4
Specify the gff file to be used for training the distributions.It should contain the position of both introns and exons
.RE
.sp
\fB\-\-bed_file\fP
.RS 4
Specify the output bedfile which contains the position of gene jumps (default: output.bed)
.RE
.sp
\fB\-\-augustus_hints\fP
.RS 4
Specify the output augustus hint file(default: augustus_hints.gff)
.RE
.SH "AUTHORS"
.sp
AUGUSTUS was written by M. Stanke, O. Keller, S. König, L. Gerischer and L. Romoth.
.SH "ADDITIONAL DOCUMENTATION"
.sp
An exhaustive documentation can be found in the file /usr/share/doc/augustus/README.md.gz.

HISTORY.TXT
     - remove bam2wig
     - compile with CGP and ZIPINPUT support by default
     - compile with SQLite and MySQL support for CGP (switch of by adding SQLite = false and MySQL = false to common.mk)
List of changes from version 3.3.2 to 3.3.3 (until Sep 13th, 2019)
     - docker container
     - new script tfix_in_frame_stop_codon_genes.py that replaces genes where spliced in-frame stop codons were predicted
     - new scripts: compare_masking.pl, merge_masking.pl
     - new program bamToWig.py as alternarive to bam2wig C++ binary
     - fix warnings on new GCC compiler (8.3)
     - introduced unit tests of C++ code (make test)
     - bugfixes (lex.cc, autoAug.pl)
     - DIAMOND as alternative to BLAST in aa2nonred.pl
List of changes from version 3.3.1 to 3.3.2 (until Oct 5th, 2018)
     - bugfixes in comparative augustus, utrrnaseq
     - new species Chiloscyllium punctatum (bamboo shark), Scyliorhinus torazame (cat shark), Rhincodon typus (whale shark)
     - updated comparative augustus (CGP) tutorial 
List of changes from version 3.3 to 3.3.1 (until May 8th, 2018)
     - new species pisaster (Pisaster ochraceus, ochre starfish)
     - bugfixes of sampling error in intron model and in check of
       frame-compatibility of CDSpart hints without frame
     - new auxiliary program utrrnaseq to find training UTRs
     - new script setStopCodonFreqs.pl
List of changes from version 3.2.3 to 3.3 (until July 11th, 2017)
     - new program ESPOCA to estimate selective pressure on codon alignments
     - gene finding on ancestral genomes is enabled
     - new default parameters for comparative gene prediction (CGP)
     - clade parameters training for CGP
     - compatibility to Ian Fiddes' Comparative Annotation Toolkit (CAT)
     - new scripts eval_dualdecomp.pl,
     - more tolerant tree parsing
     - bugfixes in augustus, joingenes, load2sqlitedb, transMap2hints.pl, splitMfasta.pl, intron2exex.pl,
       aln2wig
     - new functionality in homGeneMapping, joingenes
     - new species ciona, mnemiopsis_leidyi, nematostella_vectensis, strongylocentrotus_purpuratus
List of changes from version 3.2.2 to 3.2.3 (until December 6th, 2016)
     - bugfixes in joingenes, load2sqlitedb, transMap2hints.pl
     - added script gtf2bed.pl
     - CGP parameters for clade of 8 primates (cgp/log_reg_parameters_primates.cfg)
     - added option --clean to load2sqlitedb
     - now filter out truncated coding genes with CDS length <= 3
     - Sascha Steinbiss: remove dependency on PATH_MAX
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic license only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently

src/curve2hints.cc
/*
 * curve2hints.cc
 *
 * License: Artistic License, see file LICENSE.TXT or 
 *          https://opensource.org/licenses/artistic-license-1.0
 */

#include <iostream>
#include <string>
#include <vector>
#include <getopt.h>
#include <cstdlib>
#include "exon_seg.hh"

using namespace std;


void usage(){
  cout<<"Program to make exon segmentation and grouping the exons together to form the hints for possible position of genes to be used by augustus. \nUsage:\n curve2hints --in=file --outexons=out.gff --c=int --s=float\noptions:\n--in=s\t\tinput file is single file containing tracks in wig format\n--outexons=s\tfile for the output segmentaion\n-h\t\thelp\n--c\t\tspecify the threshold to distinguish between introns and exons in terms of coverage depths (default 4)\n--s\t\tspecify the ratio(1-lambda1)/(1-lambda2) It should be >1 (default 1.12)\n--z\t\tspecify the chunk size (default value 50000)\n--r\t\tTo enable re-estimation of distribution from the output produced\n--train_file\t\tSpecify the gff file to be used for training the distributions.It should contain the position of both introns and exons\n--bed_file\t\tSpecify the output bedfile which contains the position of gene jumps(default output.bed)\n--augustus_hints\t\tspecify the output augustus hint file(default augustus_hints.gff\n"; 
}

/* define the main file 
 */
int main(int argc, char** argv){
  int c_thresh=4,chunksize=50000;
  string inputfile=" ",outputfile=" ",train_file=" ",augustus_hints=" ",bed_file=" ";
  int c,longIndex,flag_r=0;
  float s=1.12;

  dataset coverage_info;
  static const struct option longOpts[]={
    {"in",required_argument,NULL,'i'},
    {"outexons",required_argument,NULL,'o'},
    {"train_file",required_argument,NULL,'t'},
    {"bed_file",required_argument,NULL,'b'},
    {"augustus_hints",required_argument,NULL,'a'},
    {"help",no_argument,NULL,'h'},
    {"re-estimate",no_argument,NULL,'r'},
    {"cthreshold",required_argument,NULL,'c'},
    {"sratio",required_argument,NULL,'s'},
    {"zchunksize",required_argument,NULL,'z'},
    { NULL, no_argument, NULL, 0 }
  };
  opterr = 0;
  while ((c = getopt_long (argc, argv, ":rhi:o:c:s:z",longOpts, &longIndex)) != -1){
    switch (c)
      {
      case 'i':
	inputfile = optarg;
	break;
      case 'o':
	outputfile =optarg;
	break;
      case 't':
	train_file =optarg;
	break;
      case 'b':
	bed_file =optarg;
	break;
      case 'a':
	augustus_hints =optarg;
	break;
      case 'c':
	c_thresh=atoi(optarg);
	break;
      case 'z':
	chunksize=atoi(optarg);
	break;
      case 's':
	s=atof(optarg);
	break;
      case 'r':    	
	flag_r=1;
	break;
      case 'h':    
	usage();
	exit(0);
	break;
      case 0:     /* long option without a short arg */
	cout<<" Wrong option"<<endl;
	usage();
	exit(0);
	break;
      default:
	cout<<" Wrong option"<<endl;
	usage();
	exit(0);
      } 
  }  
  if (  ((argc-optind)==0 && inputfile==" ")||  ( (argc-optind)==0 && outputfile==" ")) {
    usage();
    exit(0);
  }
  exon_segmentation seg1;
  seg1.read_file(coverage_info,inputfile.c_str());
  seg1.make_gff(coverage_info,outputfile,c_thresh,s,train_file,flag_r,bed_file,augustus_hints,chunksize);
}
